src/brod_producer.erl

```diff
@@ -247,31 +285,65 @@ terminate(_Reason, _State) ->
   ok.
 
 %%%_* Internal Functions =======================================================
+init_socket(#state{ client_pid = ClientPid
+                  , sock_mref  = OldSockMref
+                  , topic      = Topic
+                  , partition  = Partition
+                  } = State) ->
+  %% 1. Lookup, or maybe (re-)establish a connection to partition leader
+  case brod_client:get_leader_connection(ClientPid, Topic, Partition) of
+    {ok, SockPid} ->
+      ok = maybe_demonitor(OldSockMref),
+      SockMref = erlang:monitor(process, SockPid),
+      %% 2. Register self() to client.
+      ok = brod_client:register_producer(ClientPid, Topic, Partition),
 
--spec take_config(atom(), any(), config()) -> {any(), config()}.
-take_config(Key, Default, Config) when is_list(Config) ->
-  case proplists:get_value(Key, Config) of
-    ?undef -> {Default, Config};
-    Value  -> {Value, proplists:delete(Key, Config)}
+      %% 3. Update state.
+      {ok, State#state{sock_pid = SockPid, sock_mref = SockMref}};
+    {error, Error} ->
+      {error, Error}
   end.
 
--spec get_required_acks(config()) -> required_acks().
-get_required_acks(Config) when is_list(Config) ->
-  case proplists:get_value(required_acks, Config) of
-    ?undef ->
-      ?DEFAULT_REQUIRED_ACKS;
-    N when is_integer(N) ->
-      N
-  end.
+maybe_demonitor(undefined) ->
+  ok;
+maybe_demonitor(Mref) ->
+  true = erlang:demonitor(Mref, [flush]),
+  ok.
 
--spec get_ack_timeout(config()) -> pos_integer().
-get_ack_timeout(Config) when is_list(Config) ->
-  case proplists:get_value(ack_timeout, Config) of
-    ?undef ->
-      ?DEFAULT_ACK_TIMEOUT;
-    N when is_integer(N) andalso N > 0 ->
-      N
-  end.
+schedule_retry(#state{retry_tref = undefined} = State, Msg) ->
+  {ok, TRef} = timer:send_after(State#state.retry_backoff_ms, {?RETRY_MSG, Msg}),
+  {ok, State#state{retry_tref = TRef}};
+schedule_retry(State, _Msg) ->
+  %% retry timer has been already activated
+  {ok, State}.
+
+maybe_retry(#state{retries = Retries, max_retries = MaxRetries} = State0) when
+    MaxRetries =:= -1 orelse
+    Retries < MaxRetries ->
+  %% try to re-send buffered data
+  State = State0#state{retry_tref = undefined, retries = Retries + 1},
+  Buffer = State#state.buffer,
+  SockPid = State#state.sock_pid,
+  case brod_producer_buffer:maybe_send(Buffer, SockPid) of
+    {ok, Buffer} ->
+      {ok, State#state{buffer = Buffer}};
+    {error, Reason} ->
+      schedule_retry(State, Reason)
+  end;
+maybe_retry(_State) ->
+  false.
+
+is_retriable(ErrorCode) when ErrorCode =:= ?EC_CORRUPT_MESSAGE;
```

Comment:
Yes.
https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-ErrorCodes